src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -552,41 +556,141 @@ def calculate_dense(self):
     def expectation(self, state, normalize=False):
         return Hamiltonian.expectation(self, state, normalize)
 
-    def expectation_from_samples(self, freq, qubit_map=None):
-        terms = self.terms
-        for term in terms:
+    def _exp_from_circuit(
+        self, circuit: "Circuit", qubit_map: dict, nshots: int = None
+    ):
+        """
+        Calculate the expectation value from a circuit.
+        This allows for non diagonal observables. Each term of the observable is
+        treated separately, by measuring in the correct basis and re-executing the
+        circuit.
+
+        Args:
+            circuit (Circuit): input frequencies.
+            qubit_map (dict): qubit map.
+
+        Returns:
+            (float): the calculated expectation value.
+        """
+        from qibo import Circuit, gates
+
+        if nshots is None:  # pragma: no cover
+            nshots = 1000
+        rotated_circuits = []
+        coefficients = []
+        Z_observables = []
+        q_maps = []
+        for term in self.terms:
```

Comment:
Now this function is evaluating the Hamiltonian expectation value term by term. The next step (not for this PR) should be to reduce the number of circuit executions merging terms acting on non-intersecting qubit subsets.